quic/transport/ngtcp2/connection/disconnectingstate.nim

```diff
@@ -0,0 +1,51 @@
+import pkg/chronos
+import ../../../udp/datagram
+import ../../quicconnection
+import ../../connectionid
+import ../../stream
+import ./closedstate
+
+type
+  DisconnectingConnection* = ref object of ConnectionState
+    connection: QuicConnection
+    disconnect: Future[void]
+    ids: seq[ConnectionId]
+
+proc newDisconnectingConnection*(ids: seq[ConnectionId]):
+                                            DisconnectingConnection =
+  DisconnectingConnection(ids: ids)
+
+proc callDisconnect(connection: QuicConnection) {.async.} =
+  if connection.disconnect != nil:
+    await connection.disconnect()
+
+{.push locks: "unknown".}
+
+method ids*(state: DisconnectingConnection): seq[ConnectionId] =
+  state.ids
+
+method enter(state: DisconnectingConnection, connection: QuicConnection) =
+  state.connection = connection
+  state.disconnect = callDisconnect(connection)
```

Comment:
I think there should be an `assert state.connection = nil`

Reply:
maybe a `doAssert()`, this should be available everywhere not only debug builds?

Reply:
Is the purpose of the assert to ensure that `enter` is not called twice on the same state? This should hold for all states, so I could enforce it in the base `enter` call.

Reply:
I added protection against state reentry in 923ad8d.